File: src/typegraph/rust/src/runtimes/http.rs
use crate::{
    t::{self, TypeBuilder},
    wasm::{
        self,
        core::TypeId,
        runtimes::{
            BaseMaterializer, Effect, HttpMethod, HttpRuntimeData, MaterializerHttpRequest,
            RuntimeId,
        },
    },
    Result,
};

impl Default for HttpMethod {
    fn default() -> Self {
        Self::Get
    }
}

#[derive(Debug, Default)]
pub struct HttpRequestBuilder {
    runtime: RuntimeId,
    inp: TypeId,
    out: TypeId,
    method: HttpMethod,
    path: String,
    content_type: Option<String>,
    header_prefix: Option<String>,
    query_fields: Option<Vec<String>>,
    effect: Effect,
}

impl TypeBuilder for HttpRequestBuilder {
    fn build(self) -> Result<TypeId> {
        let base = BaseMaterializer {
            runtime: self.runtime,
            effect: self.effect,
        };

        let mat_options = MaterializerHttpRequest {
            method: self.method,
            path: self.path,
            content_type: self.content_type,
            header_prefix: self.header_prefix,
            query_fields: self.query_fields,
            rename_fields: None,
            body_fields: None,
            auth_token_field: None,
        };

        let mat_id = wasm::with_runtimes(|r, s| r.call_http_request(s, base, &mat_options))?;

        t::func(self.inp.build()?, self.out.build()?, mat_id)?.build()
    }
}

impl HttpRequestBuilder {
    pub fn content_type(mut self, content: impl ToString) -> Self {
        self.content_type = Some(content.to_string());
        self
    }

    pub fn header_prefix(mut self, prefix: impl ToString) -> Self {
        self.header_prefix = Some(prefix.to_string());
        self
    }

    pub fn query_fields(mut self, fields: impl IntoIterator<Item = impl ToString>) -> Self {
        self.header_prefix = Some(fields.into_iter().map(|f| f.to_string()).collect());
        self
    }

    pub fn effect(mut self, effect: Effect) -> Self {
        self.effect = effect;
        self
    }
}

#[derive(Debug)]
pub struct HttpRuntime {
    id: u32,
}

impl HttpRuntime {
    pub fn new(
        endpoint: &str,
        cert_secret: Option<&str>,
        basic_auth_secret: Option<&str>,
    ) -> Result<Self> {
        let data = HttpRuntimeData {
            endpoint: endpoint.to_string(),
            cert_secret: cert_secret.as_ref().map(|v| v.to_string()),
            basic_auth_secret: basic_auth_secret.as_ref().map(|v| v.to_string()),
        };

        let id = wasm::with_runtimes(|r, s| r.call_register_http_runtime(s, &data))?;

        Ok(Self { id })
    }

    pub fn get<I, O>(&self, inp: I, out: O) -> Result<HttpRequestBuilder>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        self.request(HttpMethod::Get, inp, out)
    }

    pub fn post<I, O>(&self, inp: I, out: O) -> Result<HttpRequestBuilder>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        self.request(HttpMethod::Post, inp, out)
            .map(|r| r.effect(Effect::Create(false)))
    }

    pub fn put<I, O>(&self, inp: I, out: O) -> Result<HttpRequestBuilder>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        self.request(HttpMethod::Put, inp, out)
            .map(|r| r.effect(Effect::Update(false)))
    }

    pub fn patch<I, O>(&self, inp: I, out: O) -> Result<HttpRequestBuilder>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        self.request(HttpMethod::Patch, inp, out)
            .map(|r| r.effect(Effect::Update(false)))
    }

    pub fn delete<I, O>(&self, inp: I, out: O) -> Result<HttpRequestBuilder>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        self.request(HttpMethod::Delete, inp, out)
            .map(|r| r.effect(Effect::Delete(false)))
    }

    fn request<I, O>(&self, method: HttpMethod, inp: I, out: O) -> Result<HttpRequestBuilder>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        Ok(HttpRequestBuilder {
            runtime: self.id,
            method,
            inp: inp.build()?,
            out: out.build()?,
            ..Default::default()
        })
    }
}

File: src/typegraph/deno/src/runtimes/http.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import * as t from "../types.ts";
import type {
  Effect,
  HttpMethod,
  MaterializerHttpRequest,
} from "../gen/runtimes.ts";
import { runtimes } from "../sdk.ts";
import { type Materializer, Runtime } from "./mod.ts";
import { fx } from "../index.ts";

type HttpRequestMat<M extends string> = Materializer &
  Omit<MaterializerHttpRequest, "method"> & {
    method: M;
  };

export class HttpRuntime extends Runtime {
  constructor(
    public endpoint: string,
    public certSecret?: string,
    public basicAuthSecret?: string,
  ) {
    super(
      runtimes.registerHttpRuntime({
        endpoint,
        certSecret,
        basicAuthSecret,
      }),
    );
  }

  #request<
    M extends HttpMethod,
    I extends t.Typedef = t.Typedef,
    O extends t.Typedef = t.Typedef,
  >(
    method: M,
    inp: I,
    out: O,
    options: Omit<MaterializerHttpRequest, "method">,
    effect: Effect,
  ): t.Func<I, O> {
    const matId = runtimes.httpRequest(
      {
        runtime: this._id,
        effect,
      },
      { method, ...options },
    );

    const mat: HttpRequestMat<M> = {
      _id: matId,
      method,
      ...options,
    };
    return t.func(inp, out, mat);
  }

  get<I extends t.Typedef = t.Typedef, O extends t.Typedef = t.Typedef>(
    inp: I,
    out: O,
    options: Omit<MaterializerHttpRequest, "method">,
  ): t.Func<I, O> {
    return this.#request("get", inp, out, options, fx.read());
  }

  post<I extends t.Typedef = t.Typedef, O extends t.Typedef = t.Typedef>(
    inp: I,
    out: O,
    options: Omit<MaterializerHttpRequest, "method">,
    effect?: Effect,
  ): t.Func<I, O> {
    return this.#request("post", inp, out, options, effect ?? fx.create());
  }

  put<I extends t.Typedef = t.Typedef, O extends t.Typedef = t.Typedef>(
    inp: I,
    out: O,
    options: Omit<MaterializerHttpRequest, "method">,
    effect?: Effect,
  ): t.Func<I, O> {
    return this.#request("put", inp, out, options, effect ?? fx.update());
  }

  patch<I extends t.Typedef = t.Typedef, O extends t.Typedef = t.Typedef>(
    inp: I,
    out: O,
    options: Omit<MaterializerHttpRequest, "method">,
    effect?: Effect,
  ): t.Func<I, O> {
    return this.#request("patch", inp, out, options, effect ?? fx.update());
  }

  delete_<I extends t.Typedef = t.Typedef, O extends t.Typedef = t.Typedef>(
    inp: I,
    out: O,
    options: Omit<MaterializerHttpRequest, "method">,
    effect?: Effect,
  ): t.Func<I, O> {
    return this.#request("delete", inp, out, options, effect ?? fx.delete_());
  }
}
